control/velocity_controller_lqr/velocity_controller_lqr/velocity_controller_lqr_lib.py

```diff
@@ -0,0 +1,289 @@
+    def __init__(
+        self,
+        q_surge,
+        q_pitch,
+        q_yaw,
+        r_surge,
+        r_pitch,
+        r_yaw,
+        i_surge,
+        i_pitch,
+        i_yaw,
+        max_force,
+    ):
+        self.integral_error_surge = 0.0
+        self.integral_error_pitch = 0.0
+        self.integral_error_yaw = 0.0
+
+        self.surge_windup = False  # Windup variable
+        self.pitch_windup = False  # Windup variable
+        self.yaw_windup = False  # Windup variable
+
+        self.max_force = max_force  # Maximum force that can be applied
+
+        self.i_surge = i_surge  # Integral gain for surge
+        self.i_pitch = i_pitch  # Integral gain for pitch
+        self.i_yaw = i_yaw  # Integral gain for yaw
+
+        self.q_surge = q_surge  # Surge cost
+        self.q_pitch = q_pitch  # Pitch cost
+        self.q_yaw = q_yaw  # Yaw cost
+
+        self.r_surge = r_surge  # Surge control cost
+        self.r_pitch = r_pitch  # Pitch control cost
+        self.r_yaw = r_yaw  # Yaw control cost
+
+        # VARIABLES
+        self.inertia_matrix_inv = np.linalg.inv(
+            np.array([[30, 0.6, 0], [0.6, 1.629, 0], [0, 0, 1.769]])
+        )  # mass matrix
+
+        self.state_weight_matrix = np.block(
+            [
+                [np.diag([self.q_surge, self.q_pitch, self.q_yaw]), np.zeros((3, 3))],
+                [np.zeros((3, 3)), np.diag([0.5, 0.5, 0.5])],
+            ]
+        )  # Augmented state cost matrix
+        self.input_weight_matrix = np.diag(
+            [self.r_surge, self.r_pitch, self.r_yaw]
+        )  # control cost matrix
+
+    @staticmethod
+    def quaternion_to_euler_angle(w: float, x: float, y: float, z: float) -> tuple:
+        """Function to convert quaternion to euler angles.
+
+        Args:
+            w: float: w component of quaternion
+            x: float: x component of quaternion
+            y: float: y component of quaternion
+            z: float: z component of quaternion
+
+        Returns:
+            X: float: roll angle
+            Y: float: pitch angle
+            Z: float: yaw angle
+
+        """
+        y_square = y * y
+
+        t0 = +2.0 * (w * x + y * z)
+        t1 = +1.0 - 2.0 * (x * x + y_square)
+        phi = np.arctan2(t0, t1)
+
+        t2 = +2.0 * (w * y - z * x)
+        t2 = +1.0 if t2 > +1.0 else t2
+        t2 = -1.0 if t2 < -1.0 else t2
+        theta = np.arcsin(t2)
+
+        t3 = +2.0 * (w * z + x * y)
+        t4 = +1.0 - 2.0 * (y_square + z * z)
+        psi = np.arctan2(t3, t4)
+
+        return phi, theta, psi
+
+    @staticmethod
+    def ssa(angle: float) -> float:
+        """Function to convert the angle to the smallest signed angle.
+
+        Args:
+            angle: float: angle in radians
+
+        Returns:
+            angle: float: angle in radians
+
+        """
+        if angle > np.pi:
+            angle -= 2 * np.pi
+        elif angle < -np.pi:
+            angle += 2 * np.pi
+        return angle
```

Comment:
why not use Fossen's ssa?

 `return (angle + np.pi) % (2 * np.pi) - np.pi`

Reply:
washed